### PR TITLE
fix: merge .percy.yml config options into serializeDOM

### PR DESCRIPTION
**Pull request by rishigupta1599**
## Summary
- `.percy.yml` config options were not being passed to `captureDOM`/`PercyDOM.serialize()` — only per-snapshot options were used
- Now merges `utils.percy.config.snapshot` as defaults with per-snapshot options taking priority

## Test plan
- [x] All 4 integration tests pass (nightwatch)
- [x] All 8 unit tests pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by rishigupta1599 — COMMENTED** [review]
Claude Code Review (automated) — 1 inline finding(s). Full report in the PR comment below. Verdict: Passed.

File: commands/percySnapshot.js [hunk 1]
```diff
@@ -47,7 +47,8 @@ module.exports = class PercySnapshotCommand {
       const readinessDiagnostics = await waitForReady(this.api, options, utils, log);
 
       // Serialize and capture the DOM
-      let { domSnapshot, url } = await captureDOM(this.api, options, utils, log, domScript);
+      const mergedOptions = utils.mergeSnapshotOptions(options);
```
Comment (rishigupta1599) [inline commands/percySnapshot.js:54]: **[Medium] Guard `mergeSnapshotOptions` for older sdk-utils**

This call is unguarded. The readiness feature in this same SDK degrades to a no-op when the helper is missing (`if (typeof utils?.runReadinessGate !== 'function') return undefined;`), but `mergeSnapshotOptions` (added in sdk-utils 1.31.15+) is not similarly guarded. On an older resolved `@percy/sdk-utils`, `utils.mergeSnapshotOptions` would be `undefined` and throw a TypeError, breaking every snapshot.

**Suggestion:** Mirror the existing pattern, e.g.
```suggestion
      const mergedOptions = typeof utils.mergeSnapshotOptions === 'function'
        ? utils.mergeSnapshotOptions(options)
        : { ...(utils?.percy?.config?.snapshot || {}), ...options };
```

_Reviewer: stack:code-review_

**Comment by rishigupta1599** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/percy-nightwatch/pull/879  •  **Head:** e91af10  •  **Reviewers:** stack:code-review

## Summary
Merges global `.percy.yml` snapshot config (`utils.percy.config.snapshot`) with per-snapshot `options` — per-call options taking priority — and passes the merged result into `captureDOM`/`serializeDOM`, so config-level serialization options (e.g. responsive/JS options) are honored during DOM capture (PER-8053).

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | N/A | Security lens out of scope for this review. |
| High | Security | Authentication/authorization checks present | N/A | Security lens out of scope for this review. |
| High | Security | Input validation and sanitization | N/A | Security lens out of scope for this review. |
| High | Security | No IDOR — resource ownership validated | N/A | Security lens out of scope for this review. |
| High | Security | No SQL injection (parameterized queries) | N/A | Security lens out of scope for this review. |
| High | Correctness | Logic is correct, handles edge cases | Pass | `mergeSnapshotOptions` does `{...configOptions, ...options}` so per-call wins; merged options correctly flow into `captureDOM` at `commands/percySnapshot.js:51`. `postData` intentionally keeps raw `options` so config defaults are not re-posted to the CLI (CLI applies config itself) — consistent. |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | Call sits inside the existing try/catch; no new swallowed paths. |
| High | Correctness | No race conditions or concurrency issues | Pass | Synchronous pure merge, no shared mutable state. |
| Medium | Testing | New code has corresponding tests | Fail | No test in this repo exercises the merge path in `percySnapshot` (config option present + per-call override). `mergeSnapshotOptions` is unit-tested only in `@percy/sdk-utils`. See finding. |
| Medium | Testing | Error paths and edge cases tested | Fail | No test for the older-sdk-utils case where `utils.mergeSnapshotOptions` is undefined. See finding. |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | PR reports 4 integration + 8 unit tests pass; change is additive to the capture path. |
| Medium | Performance | No N+1 queries or unbounded data fetching | Pass | Single in-memory object spread; negligible. |
| Medium | Performance | Long-running tasks use background jobs | N/A | No long-running work introduced. |
| Medium | Quality | Follows existing codebase patterns | Pass | Reuses shared `utils.mergeSnapshotOptions` helper rather than re-implementing; matches the SDK's util-delegation pattern. |
| Medium | Quality | Changes are focused (single concern) | Pass | One-line behavioral change scoped to option merging before capture. |
| Low | Quality | Meaningful names, no dead code | Pass | `mergedOptions` is clear; no dead code. |
| Low | Quality | Comments explain why, not what | Pass | Helper carries explanatory comments in sdk-utils; call site is self-evident. |
| Low | Quality | No unnecessary dependencies added | Pass | No new deps; relies on already-required `@percy/sdk-utils` (`^1.31.15-beta.0`). |

## Findings

- **File:** `commands/percySnapshot.js:50`
- **Severity:** Medium
- **Reviewer:** stack:code-review
- **Issue:** The new `utils.mergeSnapshotOptions(options)` call is not guarded for older `@percy/sdk-utils` versions. The adjacent readiness feature added in this SDK explicitly degrades to a no-op (`if (typeof utils?.runReadinessGate !== 'function') return undefined;` in `lib/snapshot.js:78`) precisely so an older CLI/utils install does not break. `mergeSnapshotOptions` was added in the same util line (1.31.15+); on an older resolved utils it would be `undefined` and throw a `TypeError`, breaking every snapshot. The declared range `^1.31.15-beta.0` reduces but does not eliminate this in transitive/hoisted installs.
- **Suggestion:** Mirror the existing pattern, e.g. `const mergedOptions = typeof utils.mergeSnapshotOptions === 'function' ? utils.mergeSnapshotOptions(options) : { ...(utils?.percy?.config?.snapshot || {}), ...options };` (or fall back to raw `options`).

- **File:** `commands/percySnapshot.js:50`
- **Severity:** Medium
- **Reviewer:** stack:code-review
- **Issue:** No test in `percy-nightwatch` covers the merge behavior at the command level — that a `.percy.yml` config option reaches `captureDOM` and that a per-call option overrides it. The helper is unit-tested in sdk-utils, but the integration into this command (the actual fix) is uncovered.
- **Suggestion:** Add a unit test in `test/unit/snapshot.test.js` (or the command test) that stubs `utils.percy.config.snapshot` plus a per-call option and asserts the value passed into `captureDOM`/serialize reflects the correct precedence.

---

**Verdict: PASS**

**Review by pranavz28 — APPROVED** [review]
